src/openforms/submissions/attachments.py

```diff
@@ -0,0 +1,211 @@
+
+import PIL
+from glom import glom
+from PIL import Image
+
+from openforms.submissions.models import (
+    Submission,
+    SubmissionFileAttachment,
+    SubmissionStep,
+    TemporaryFileUpload,
+)
+
+DEFAULT_IMAGE_MAX_SIZE = (10000, 10000)
+
+
+def temporary_upload_uuid_from_url(url: str) -> Optional[str]:
+    try:
+        match = resolve(urlparse(url)[2])
+    except Resolver404:
+        return None
+    else:
+        if match.view_name == "api:submissions:temporary-file":
+            return str(match.kwargs["uuid"])
+        else:
+            return None
+
+
+def temporary_upload_from_url(url: str) -> Optional[TemporaryFileUpload]:
+    uuid = temporary_upload_uuid_from_url(url)
+    if not uuid:
+        return None
+    try:
+        return TemporaryFileUpload.objects.get(uuid=uuid)
+    except TemporaryFileUpload.DoesNotExist:
+        return None
+
+
+def clean_mime_type(mimetype: str) -> str:
+    # cleanup a user supplied mime-type against injection attacks
+    m = re.match(r"^[a-z0-9\.+-]+\/[a-z0-9\.+-]+", mimetype)
+    if not m:
+        return "application/octet-stream"
+    else:
+        return m.group()
+
+
+def append_file_num_postfix(
+    original_name: str, new_name: str, num: int, max: int
+) -> str:
+    if not new_name:
+        return ""
+
+    _, ext = os.path.splitext(original_name)
+    new_name, _ = os.path.splitext(new_name)
+    if max <= 1:
+        postfix = ""
+    else:
+        digits = len(str(max))
+        postfix = f"-{num:0{digits}}"
+    return f"{new_name}{postfix}{ext}"
+
+
+def attach_uploads_to_submission_step(submission_step: SubmissionStep) -> list:
+    # circular import
+    from openforms.registrations.tasks import resize_submission_attachment
+
+    components = list(submission_step.form_step.iter_components(recursive=True))
+
+    uploads = resolve_uploads_from_data(components, submission_step.data)
+
+    result = list()
+    for key, (component, uploads) in uploads.items():
+        # grab resize settings
+        resize_apply = glom(component, "image.resize.apply", default=False)
+        resize_size = (
+            glom(component, "image.resize.width", default=DEFAULT_IMAGE_MAX_SIZE[0]),
+            glom(component, "image.resize.height", default=DEFAULT_IMAGE_MAX_SIZE[1]),
+        )
+
+        base_name = glom(component, "file.name", default="")
+
+        # formio sends a list of uploads even with multiple=False
+        for i, upload in enumerate(uploads, start=1):
+            file_name = append_file_num_postfix(
+                upload.file_name, base_name, i, len(uploads)
+            )
+
+            attachment, created = SubmissionFileAttachment.objects.create_from_upload(
+                submission_step, key, upload, file_name=file_name
+            )
+            result.append((attachment, created))
+
+            if created and resize_apply and resize_size:
+                resize_submission_attachment.delay(attachment.id, resize_size)
```

Comment:
There's a possible race condition here where the image resizing has not completed before the submission is completed (and registration kicks off). This could lead to the original images without resizing being uploaded into a zaaksysteem/documenten API.

Given the user interaction required I don't think this will occur, but it's technically possible. I've created #507 so that it's at least on the backlog. Please update the code with a comment pointing to the possible problem and include a link to #507.

Reply:
We'll probaly have to re-evaluate the async/completion handling as the payment feature will add complications to it as well.